# Asaph
Asaph enables analysis of large inversions from SNP data.  Asaph employs streaming and dimensionality reduction to be fast and maintain a small memory footprint.  Asaph is developed primarily by [RJ Nowling](http://rnowling.github.io/) and students at [MSOE](https://msoe.edu).

Build status: [![Build Status](https://travis-ci.com/rnowling/asaph.svg?branch=main)](https://travis-ci.com/rnowling/asaph)

## Using Asaph

Tutorials for using Asaph for analyses described in our papers are located in the `tutorials` directory.  A few quick links:

* [Installing Asaph](tutorials/installing-asaph.md)
* [Testing Changes to Asaph](tutorials/testing-asaph.md) (mostly for developers)
* [Detecting Inversions](tutorials/detecting-inversions-with-pca.md)

## Citing Asaph
Please cite the following papers if you are using Asaph in your work:

* RJ Nowling, KR Manke, and SJ Emrich. [Detecting inversions with PCA in the presence of population structure](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0240429). *PLOS One* 2020.
* RJ Nowling and SJ Emrich. [Detecting Chromosomal Inversions from Dense SNPs by Combining PCA and Association Tests](/publications/ACMBCB_2018.pdf). *Proceedings of the 9th ACM International Conference on Bioinformatics, Computational Biology and Health Informatics* (ACM-BCB 2018), August 2018.
* RJ Nowling and SJ Emrich. [Adjusted Likelihood-Ratio Test for Variants with Unknown Genotypes](https://www.worldscientific.com/doi/10.1142/S0219720018400206). *Journal of Bioinformatics and Computational Biology* (JBCB), 16(5) 2018.


